File: src/main/java/pl/mjedynak/idea/plugins/builder/psi/MethodCreator.java
package pl.mjedynak.idea.plugins.builder.psi;

import com.intellij.psi.PsiElementFactory;
import com.intellij.psi.PsiField;
import com.intellij.psi.PsiMethod;
import org.jetbrains.annotations.NotNull;
import pl.mjedynak.idea.plugins.builder.settings.CodeStyleSettings;

public class MethodCreator {

    private final PsiElementFactory elementFactory;
    private final String builderClassName;

    public MethodCreator(PsiElementFactory elementFactory, String builderClassName) {
        this.elementFactory = elementFactory;
        this.builderClassName = builderClassName;
    }

    public PsiMethod createMethod(
            @NotNull PsiField psiField, String methodPrefix, String srcClassFieldName, boolean useSingleField) {
        String fieldName = psiField.getName();
        String fieldType = psiField.getType().getPresentableText();
        String fieldNamePrefix = CodeStyleSettings.FIELD_NAME_PREFIX;
        String fieldNameWithoutPrefix = fieldName.replaceFirst(fieldNamePrefix, "");
        String parameterNamePrefix = CodeStyleSettings.PARAMETER_NAME_PREFIX;
        String parameterName = parameterNamePrefix + fieldNameWithoutPrefix;
        String methodName = MethodNameCreator.createMethodName(methodPrefix, fieldNameWithoutPrefix);
        String methodText;
        if (useSingleField) {
            String setterName = MethodNameCreator.createMethodName("set", fieldNameWithoutPrefix);
            methodText = "public " + builderClassName + " " + methodName + "(" + fieldType + " " + parameterName
                    + ") { " + srcClassFieldName + "." + setterName + "(" + fieldName + "); return this; }";
        } else {
            methodText = "public " + builderClassName + " " + methodName + "(" + fieldType + " " + parameterName
                    + ") { this." + fieldName + " = " + parameterName + "; return this; }";
        }
        return elementFactory.createMethodFromText(methodText, psiField);
    }
}

File: src/main/kotlin/io/github/kingg22/kobra/builder/psi/MethodCreator.kt
package io.github.kingg22.kobra.builder.psi

import com.intellij.psi.PsiElementFactory
import com.intellij.psi.PsiField
import com.intellij.psi.PsiMethod
import io.github.kingg22.kobra.builder.settings.CodeStyleSettings

// TODO convert in helper util class
public class MethodCreator(private val elementFactory: PsiElementFactory, private val builderClassName: String?) {
    public fun createMethod(
        psiField: PsiField,
        methodPrefix: String?,
        srcClassFieldName: String?,
        useSingleField: Boolean,
    ): PsiMethod {
        val fieldName = psiField.name
        val fieldType = psiField.type.presentableText
        val fieldNamePrefix: String = CodeStyleSettings.FIELD_NAME_PREFIX
        val fieldNameWithoutPrefix = fieldName.replaceFirst(fieldNamePrefix.toRegex(), "")
        val parameterNamePrefix: String = CodeStyleSettings.PARAMETER_NAME_PREFIX
        val parameterName = parameterNamePrefix + fieldNameWithoutPrefix
        val methodName = MethodNameCreator.createMethodName(methodPrefix, fieldNameWithoutPrefix)
        val methodText = if (useSingleField) {
            val setterName = MethodNameCreator.createMethodName("set", fieldNameWithoutPrefix)
            "public $builderClassName $methodName($fieldType $parameterName) { $srcClassFieldName.$setterName($fieldName); return this; }"
        } else {
            "public $builderClassName $methodName($fieldType $parameterName) { this.$fieldName = $parameterName; return this; }"
        }
        return elementFactory.createMethodFromText(methodText, psiField)
    }
}
